google/cloud/pubsub/internal/subscriber_connection_impl.cc

```diff
@@ -55,16 +55,22 @@ StatusOr<pubsub::PullResponse> SubscriberConnectionImpl::Pull() {
   request.set_subscription(subscription.FullName());
   request.set_max_messages(1);
 
-  auto response = google::cloud::internal::RetryLoop(
-      current.get<pubsub::RetryPolicyOption>()->clone(),
-      current.get<pubsub::BackoffPolicyOption>()->clone(),
-      google::cloud::Idempotency::kIdempotent,
-      [stub = stub_](auto& context, auto const& request) {
-        return stub->Pull(context, request);
-      },
-      request, __func__);
-  if (!response) return std::move(response).status();
-  if (response->received_messages_size() != 1) {
+  auto messages_received = 0;
+  StatusOr<google::pubsub::v1::PullResponse> response;
+  while (messages_received == 0) {
+    response = google::cloud::internal::RetryLoop(
+        current.get<pubsub::RetryPolicyOption>()->clone(),
+        current.get<pubsub::BackoffPolicyOption>()->clone(),
+        google::cloud::Idempotency::kIdempotent,
+        [stub = stub_](auto& context, auto const& request) {
+          return stub->Pull(context, request);
+        },
+        request, __func__);
+
+    if (!response) return std::move(response).status();
+    messages_received = response->received_messages_size();
+  }
```

Comment:
This implies we have a retry loop inside a retry loop.  That seems odd.  If I (or an application developer writes):

```cc
  auto m = client.Pull(..., Options{}.set<pubsub::RetryPolicyOption>(LimitedTimeRetryPolicy(60s)));
```

I expect the client library to block for at most 60s, not for "60s unless no message arrive in which case block for 60s more"?

Maybe something like this instead?

```suggestion
    auto response = google::cloud::internal::RetryLoop(
        current.get<pubsub::RetryPolicyOption>()->clone(),
        current.get<pubsub::BackoffPolicyOption>()->clone(),
        google::cloud::Idempotency::kIdempotent,
        [stub = stub_](auto& context, auto const& request) {
          auto r = stub->Pull(context, request);
          if (!r || !r->received_messages().empty()) return r;
          // No messages received, return a retryable error to
          // try again. This respects the retry policy.
          return google::cloud::internal::AbortedError("no messages received", GCP_ERROR_INFO());
        },
        request, __func__);
```

Reply:
I am not sure `AbortedError` is the right way to represent this condition.  We may need a different error code, or a completely different retry loop.  But the point about the double loop stands I think.

Reply:
Added a custom retry loop.